tools/hls-verifier/src/HlsVhdlTb.cpp

```diff
@@ -626,188 +531,127 @@ string HlsVhdlTb::getMemoryInstanceGeneration() {
-  code << "\t\tport map (" << endl;
-  for (size_t i = 0; i < duvPortMap.size(); i++) {
-    pair<string, string> elem = duvPortMap[i];
-    code << "\t\t\t" << elem.first << " => " << elem.second
-         << ((i < duvPortMap.size() - 1) ? "," : "") << endl;
-  }
-  code << "\t\t);" << endl << endl;
-  return code.str();
+  duvInst.emitVhdl(os);
 }
 
-string HlsVhdlTb::getCommonBody() { return commonTbBody; }
-
-string HlsVhdlTb::getArchitectureEnd() { return "end architecture behav;\n"; }
-
-string HlsVhdlTb::getOutputTagGeneration() {
-  stringstream out;
-  out << "\n";
-  out << "---------------------------------------------------------------------"
-         "-------\n";
-  out << "-- Write \"[[[runtime]]]\" and \"[[[/runtime]]]\" for output "
-         "transactor\n";
+void HlsVhdlTb::getOutputTagGeneration(mlir::raw_indented_ostream &os) {
+  os << "-- Write [[[runtime]]], [[[/runtime]]] for output transactor\n";
   for (auto &cDuvParam : cDuvParams) {
     if (cDuvParam.isOutput) {
-      out << "write_output_transactor_" << cDuvParam.parameterName
-          << "_runtime_proc : process\n";
-      out << "	file fp             : TEXT;\n";
-      out << "	variable fstatus    : FILE_OPEN_STATUS;\n";
-      out << "	variable token_line : LINE;\n";
-      out << "	variable token      : STRING(1 to 1024);\n";
-      out << "\n";
-      out << "begin\n";
-      out << "	file_open(fstatus, fp, OUTPUT_" << cDuvParam.parameterName
-          << ", WRITE_MODE);\n";
-      out << "	if (fstatus /= OPEN_OK) then\n";
-      out << "		assert false report \"Open file \" & OUTPUT_"
-          << cDuvParam.parameterName << " & \" failed!!!\" severity note;\n";
-      out << "		assert false report \"ERROR: Simulation using HLS TB "
-             "failed.\" severity failure;\n";
-      out << "	end if;\n";
-      out << "	write(token_line, string'(\"[[[runtime]]]\"));\n";
-      out << "	writeline(fp, token_line);\n";
-      out << "	file_close(fp);\n";
-      out << "	while transaction_idx /= TRANSACTION_NUM loop\n";
-      out << "		wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	end loop;\n";
-      out << "	wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	file_open(fstatus, fp, OUTPUT_" << cDuvParam.parameterName
-          << ", APPEND_MODE);\n";
-      out << "	if (fstatus /= OPEN_OK) then\n";
-      out << "		assert false report \"Open file \" & OUTPUT_"
-          << cDuvParam.parameterName << " & \" failed!!!\" severity note;\n";
-      out << "		assert false report \"ERROR: Simulation using HLS TB "
-             "failed.\" severity failure;\n";
-      out << "	end if;\n";
-      out << "	write(token_line, string'(\"[[[/runtime]]]\"));\n";
-      out << "	writeline(fp, token_line);\n";
-      out << "	file_close(fp);\n";
-      out << "	wait;\n";
-      out << "end process;\n";
+
+      os << llvm::formatv(procWriteTransactions, cDuvParam.parameterName,
```

Comment:
I think you can repeat `{0}` multiple times, and you only need to pass `cDuvParam.parameterName` once. https://llvm.org/docs/ProgrammersManual.html#formatting-strings-the-formatv-function